### PR TITLE
EF-4826 - audit user role changes

### PR DESCRIPTION
**Pull request by djacob-pd**
**Link back to Jira ticket:** https://pagerduty.atlassian.net/browse/EF-4826

## Description

A script that uses the Audit Trail API to retrieve audit records from the Global Admin endpoint and filter it to show just user role changes.

### Steps to test changes
* create come users
* change their role, and role tiers
* use the docs to run the script

## Documentation 
- [ ] There is documentation that needs to be updated upon merging these changes

Links to any documentation to be updated:

### CODE REVIEW

**Review by johncoleman83 — APPROVED** [review]

File: get_info_on_all_users/README.md [hunk 1]
```diff
@@ -57,12 +57,87 @@ Running the script requires the provision of just one argument: a global REST AP
...
+#### Write Results to CSV
+
+```
+./get_user_role_changes.py -k API-KEY-HERE --filename user_role_changes.csv
+```
+
+#### Show all user change audit records
+
+In some cases you may want to see all user changes or the complete role change audit record.
+This option provides a basic implementation to print the reocrds in JSON format.
```
Comment (johncoleman83) [inline get_info_on_all_users/README.md:123]:
```suggestion
This option provides a basic implementation to print the records in JSON format.
```

**Review by rafusel — APPROVED** [review]

File: get_info_on_all_users/README.md [hunk 2]
```diff
@@ -103,6 +103,16 @@ Running the script requires one argument: a global REST API key
 ./get_user_role_changes.py -k API-KEY-HERE --user-id PABC123
 ```
 
+#### Exclude user create and deletes
+
+By default when a user is create this script will include role changes from user creates and deletes,
```
Comment (johncoleman83) [inline get_info_on_all_users/README.md:108]:
```suggestion
By default this script will include role changes from user creates and deletes
```

File: get_info_on_all_users/get_user_role_changes.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        print('Could not get user role change audit records')
+        raise e
+
+if __name__ == '__main__':
+    ap = argparse.ArgumentParser(description='Prints all user role or tier changes between the given dates')
+    ap.add_argument('-k', '--api-key', required=True, help='REST API key')
+    ap.add_argument('-s', '--since', required=False, help='Start of date range to search')
+    ap.add_argument('-u', '--until', required=False, help='End of date range to search')
+    ap.add_argument('-i', '--user-id', required=False, help='Filter results to a single user ID')
+    ap.add_argument('-o', '--only-updates', action='store_true', help='Exclude user creates and deletes from role change results')
```
Comment (johncoleman83) [inline get_info_on_all_users/get_user_role_changes.py:184]: I like this option, sounds like this was the original request for this script.

I am glad though were still keeping the create / delete changes. I've seen some fraud activity where users are created and deleted too and I could see why a customer would want to get all the user changes including creates and deletes especially if the user was never actually updated.

File: get_info_on_all_users/README.md [hunk 4]
```diff
@@ -57,12 +57,98 @@ Running the script requires the provision of just one argument: a global REST AP
...
+#### Filter by User ID
+
+```
+./get_user_role_changes.py -k API-KEY-HERE --user-id PABC123
+```
+
+#### Exclude user create and deletes
+
+By default this script will include role changes from user creates and deletes
+(None -> New Role, Old Role -> None). To exlucde these from reported results use the `--only-updates`
```
Comment (johncoleman83) [inline get_info_on_all_users/README.md:109]:
```suggestion
(None -> New Role, Old Role -> None). To exclude these from reported results use the `--only-updates`
```

oops another type-o

**Comment by ethanmsl** [comment]
Minor bug.
Impact:
Appending a full csv *including* header row will make using the csv difficult.
The csv will be syntactically valid, but semantically broken.
(I'd just remove the appending altogether for simplicity of use and code -- but appending without header good too, ofc.)

#Prob: 
### **get_users_by_role** appends its header each time it runs 


```sh
python get_users_by_role.py -k u+wsecretsJ6secretsQ  -r=owner -f bap.csv
python get_users_by_role.py -k u+wsecretsJ6secretsQ  -r=admin -f bap.csv
```
-->
```csv
Name,Role,Email                                              <=========
EMSL,owner,ethanmsl@gmail.com
Name,Role,Email                                              <=========
Ethan Skowronski-Lutz,admin,eskowronski-lutz@pagerduty.com
King Aragorn,admin,strider@example.com
Muuuuta Biiiility,admin,125.greenholt.earline@graham.name
```

# Fix
### (A) either: remove the `'a+'` append designator here.  (multiple roles can already be specified in one run)
```python
def write_rows(column1, column2, column3):
# one function for writing to csv
  with open(filename, 'a+') as csvfile:  # <================
    writer = csv.writer(csvfile)
    writer.writerow([column1, column2, column3])
```
### (B) or: do a check for a blank file here
```python
if __name__ == '__main__':
  ...
  if args.filename[-4:] != '.csv':
    filename = args.filename + '.csv'
  else:
    filename = args.filename
  write_rows('Name','Role', 'Email')  # <=============================
  for role in roles:
   ...
flag:\nteam_managers\n")
      for api_value in allowed_roles:
        sys.stdout.write(api_value+"\n")
      sys.stdout.write("\n")  
  for role_type, total in role_types_count.items():
    sys.stdout.write(role_type+": "+str(total)+"\n")
```

**Pull request merged**